src/gpu/align_z_to_other.cu

```diff
@@ -0,0 +1,239 @@
+//
+// Created by konrad on 10/23/17.
+//
+
+#include "../../include/librealsense/gpu/align_z_to_other_helpers.h"
+#include "../../include/librealsense/rs.h"
+#include "../types.h"
+#include <iostream>
+#include <cmath>
+#include <mutex>
+
+#ifndef DEBUG
+#define DEBUG(x) do {} while(0);
+#else
+#define DEBUG(x) do { std::cerr << x << std::endl; } while(0);
+#endif
+
+#if !defined (HAVE_CUDA) || defined (CUDA_DISABLER)
+
+
+
+namespace gpu {
+bool align_z_to_other(rsimpl::byte * z_aligned_to_other, const uint16_t * z_pixels, float z_scale, const rs_intrinsics & z_intrin, const rs_extrinsics & z_to_other, const rs_intrinsics & other_intrin) {
+    DEBUG("GPU DISABLED");
+    return false;
+}
+}
+
+#else
+namespace gpu {
+
+    void __global__ cast_to_ushort(unsigned short * output, unsigned int * input, const rs_intrinsics intrin) {
+        int x = blockIdx.x * blockDim.x + threadIdx.x;
+        int y = blockIdx.y * blockDim.y + threadIdx.y;
+
+        if (x >= intrin.width || y >= intrin.height) {
+            return;
+        }
+
+        int pixel_index = y * intrin.width + x;
+        output[pixel_index] = static_cast<unsigned short>(input[pixel_index]);
+
+    }
+
+    __device__ inline float get_depth(int z_pixel_index, float z_scale, unsigned short * z_pixels_gpu) { return z_scale * z_pixels_gpu[z_pixel_index]; }
+
+    __device__ inline void transfer_pixel(int z_pixel_index, int other_pixel_index, unsigned short * z_pixels_gpu, unsigned int * out_z_gpu) {
+        if(0 != atomicCAS(&(out_z_gpu[other_pixel_index]), 0, (unsigned int) z_pixels_gpu[z_pixel_index])) {
+            atomicMin(&(out_z_gpu[other_pixel_index]), (unsigned int) z_pixels_gpu[z_pixel_index]);
+        }
+    }
+
+void __global__ align_images(const rs_intrinsics depth_intrin, const rs_extrinsics depth_to_other,
+                      const rs_intrinsics other_intrin, float z_scale, unsigned short * z_pixels_gpu, unsigned int * out_pixels_gpu) {
+
+
+
+        float inv_fx = 1.0 / depth_intrin.fx;
+        float inv_fy = 1.0 / depth_intrin.fy;
+        // Iterate over the pixels of the depth image
+
+        int depth_x = blockIdx.x * blockDim.x + threadIdx.x;
+        int depth_y = blockIdx.y * blockDim.y + threadIdx.y;
+
+        if (depth_x >= depth_intrin.width || depth_y >= depth_intrin.height) {
+            return;
+        }
+
+        int depth_pixel_index = depth_y * depth_intrin.width + depth_x;
+
+        // Skip over depth pixels with the value of zero, we have no depth data so we will not write anything into our aligned images
+        if (float depth = get_depth(depth_pixel_index, z_scale, z_pixels_gpu)) {
```

Comment:
NOTE: I prefer this syntax when it's practical, but I believe the current standard is to declare variables outside of if/while/switch statements. We should discuss this with the group and come to a decision (especially with the increased support for this feature in c++14)